### PR TITLE
ExtendedClient Update

### PR DESCRIPTION
**Pull request by bowenjw**
## Changes to ExtendedClient

- Changed require and import
  - As part of this `default` export method to event, command, and interaction files
  - Added error handling for a directory being not found
- Remove the init method 
  - added constructor adds items to commands and interactions collections
  - added login method checks key is not undefined and runs `super.login`
- Made `feature/deploy.ts` a public method of ExtendedClient
- Split the interaction collection into three collections for modals, select menus, and buttons
- Made the collections and config properties `readonly` from `public`

## interactionCreate event
- Updated to exclusively switch statement. Uses fall though for for types in the same collection for example
```ts
switch (interaction.commandType) {
// Context Menu
case ApplicationCommandType.Message:
case ApplicationCommandType.User:
	client.contextMenus.get(interaction.commandName)?.execute(client, interaction);
	break;
default:
	break;
}
```
- Try block covers the whole switch statement and can send error message if interaction is deferred

## Interfaces
Command interfaces have been grouped together in `src\interfaces\Command.ts` and are now extended from on core interface. The extended interfaces are as follows:
- ChatInputCommand
- ContextMenu
  - UserContextMenu
  - MessageContextMenu

Command interfaces have been grouped together in `src\interfaces\Interaction.ts` and are now extended from on core interface. The extended interfaces are as follows:
- Button
- ModalSubmit
- AnySelectMenu
  - StringSelectMenu
  - MentionableSelectMenu
  - ChannelSelectMenu
  - RoleSelectMenu
  - UserSelectMenu

## Other Changes
- Added new example command `/ping`. Replies with pong and a ping button that replies with pong when pressed
    - File location `src\commands\example\ping.ts`, `src\features\ping.ts`, and `src\interactions\buttons\example\ping.ts`
- Added [ESLint](https://eslint.org/) with Type script extensions using the ruls from the [discord.js guide](https://discordjs.guide/preparations/setting-up-a-linter.html#setting-up-eslint-rules). 
_Note: The discord.js ESLint file has indent set to tab it has been changed it to 4 spaces._ 
- Added [Nodemon](https://nodemon.io/) and updated the dev script to `nodemon src/bot.ts --no-deployment`
- renamed `.env.example` to `example.env`

### CODE REVIEW

File: src/events/interactionCreate.ts [hunk 1]
```diff
@@ -1,95 +1,88 @@
...
-                try {
-                    await modal!.execute(client, interaction);
-                } catch (error) {
-                    if (error! instanceof Error) {
-                        console.error(error);
-                        await replyError(interaction);
-                    }
-                }
-
+                interactionName = interaction.customId.split(' ')[0];
```
Comment (ppauel) [inline src/events/interactionCreate.ts:73]:
```
interactionName = client.config.interactions.splitCustomId ? interaction.customId.split('_')[0] : interaction.customId;
```

**Review by ppauel — APPROVED** [review]
Very helpful and type-safe changes. Also thank you for the detailed summary, I really appreciate your work, John! 🙌

File: src/classes/Client.ts [hunk 2]
```diff
@@ -1,84 +1,176 @@
...
+        console.log('Starting up...');
 
         // Paths
         const commandPath = path.join(__dirname, '..', 'commands'),
             contextMenuPath = path.join(__dirname, '..', 'context_menus'),
-            interactionPath = path.join(__dirname, '..', 'interactions'),
+            buttonPath = path.join(__dirname, '..', 'interactions', 'buttons'),
+            selectMenuPath = path.join(__dirname, '..', 'interactions', 'select_menus'),
+            modalPath = path.join(__dirname, '..', 'interactions', 'modals'),
             eventPath = path.join(__dirname, '..', 'events');
```
Comment (ppauel) [inline src/classes/Client.ts:50]: fixed folder structure https://github.com/ppauel/typescript-discord-bot/commit/52759f953b1fad4d8dfb594aad38c19fe5c88bb7

**Comment by bowenjw** [comment]
> Very helpful and type-safe changes. Also thank you for the detailed summary, I really appreciate your work, John! 🙌

Glad I could help

**Pull request merged**